cmd/cryptography/checksum.go

```diff
@@ -0,0 +1,246 @@
+		base = filepath.Clean(destinationDir) + string(filepath.Separator) + strings.TrimSuffix(string(filepath.Base(sourcePack)), ".pack")
+	}
+	checksumFilename := base + "." + strings.Replace(hashFunction, "-", "", -1) + ".checksum"
+
+	if utils.FileExists(checksumFilename) {
+		log.Errorf("\"%s\" already exists, choose a diferent path", checksumFilename)
+		return errs.ErrPathAlreadyExists
+	}
+
+	// Unpack it to the same directory as the .pack
+	packDir := strings.TrimSuffix(sourcePack, filepath.Ext(sourcePack)) + string(os.PathSeparator)
+	if utils.DirExists(packDir) {
+		log.Errorf("pack was already extracted to \"%s\"", packDir)
+		return errs.ErrPathAlreadyExists
+	}
+
+	zipReader, err := zip.OpenReader(sourcePack)
+	if err != nil {
+		log.Errorf("can't decompress \"%s\": %s", sourcePack, err)
+		return errs.ErrFailedDecompressingFile
+	}
+	var packFileList []string
+	for _, file := range zipReader.File {
+		if err := utils.SecureInflateFile(file, packDir, ""); err != nil {
+			if err == errs.ErrTerminatedByUser {
+				log.Infof("aborting pack extraction. Removing \"%s\"", packDir)
+				return os.RemoveAll(packDir)
+			}
+			return err
+		}
+		packFileList = append(packFileList, filepath.Join(filepath.Clean(packDir), filepath.Clean(file.Name)))
+	}
+
+	digests, err := getChecksumList(packDir, packFileList, hashFunction)
+	if err != nil {
+		return err
+	}
+
+	out, err := os.Create(checksumFilename)
+	if err != nil {
+		log.Error(err)
+		return errs.ErrFailedCreatingFile
+	}
+	defer out.Close()
+	for i := 0; i < len(digests); i++ {
+		_, err := out.Write([]byte(digests[i].digest + " " + digests[i].filename + "\n"))
+		if err != nil {
+			return err
+		}
+	}
+	// Cleanup extracted pack
+	log.Debugf("deleting \"%s\"", packDir)
+	if err := os.RemoveAll(packDir); err != nil {
+		return err
+	}
+	return nil
+}
+
+func VerifyChecksum(sourcePack, checksum string) error {
+	if !utils.FileExists(sourcePack) {
+		log.Errorf("\"%s\" is not a valid .pack", sourcePack)
+		return errs.ErrFileNotFound
+	}
+	if !utils.FileExists(checksum) {
+		log.Errorf("\"%s\" does not exist", checksum)
+		return errs.ErrFileNotFound
+	}
+	hashAlgorithm := filepath.Ext(strings.Split(checksum, ".checksum")[0])[1:]
+	if !isValidHash(hashAlgorithm) {
+		log.Errorf("\"%s\" is not a valid .checksum file (correct format is [<pack>].[<hash-algorithm>].checksum). Please confirm if the algorithm is supported.", checksum)
+		return errs.ErrInvalidHashFunction
+	}
+
+	// Unpack it to the same directory as the .pack
+	packDir := strings.TrimSuffix(sourcePack, filepath.Ext(sourcePack)) + string(os.PathSeparator)
+	if utils.DirExists(packDir) {
+		log.Errorf("pack was already extracted to \"%s\"", packDir)
+		return errs.ErrPathAlreadyExists
+	}
+
+	zipReader, err := zip.OpenReader(sourcePack)
+	if err != nil {
+		log.Errorf("can't decompress \"%s\": %s", sourcePack, err)
+		return errs.ErrFailedDecompressingFile
+	}
+	var packFileList []string
+	for _, file := range zipReader.File {
```

Comment:
Decompression can be avoided by using the same logic as I commented above: https://github.com/Open-CMSIS-Pack/cpackget/pull/108/commits/57f2022f27d8fe680b92798c3ae6a816669043e9#r923441717

But now I see that you also check if there are files missing or extra in the .pack file. This can be done by using the digests in the checksum file in a `map`, then you can tell whether or not there are any extra or missing files in the offending .pack

Reply:
Done!